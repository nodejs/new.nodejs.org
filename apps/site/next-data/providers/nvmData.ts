import { cache } from 'react';

import generateNvmData from '@/next-data/generators/nvmData.mjs';

const nvmData = await generateNvmData();

export default cache(() => nvmData);
