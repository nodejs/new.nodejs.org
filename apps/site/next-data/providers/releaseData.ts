import { cache } from 'react';

import generateReleaseData from '@/next-data/generators/releaseData.mjs';

const provideReleaseData = cache(() => generateReleaseData());

export default provideReleaseData;
