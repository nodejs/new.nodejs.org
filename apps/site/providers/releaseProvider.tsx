'use client';

import type { Dispatch, PropsWithChildren, FC } from 'react';
import { createContext, useMemo, useReducer } from 'react';

import type { NodeRelease } from '@/types';
import type {
  ReleaseDispatchActions,
  ReleaseAction,
  ReleaseContextType,
  ReleaseProviderProps,
  ReleaseState,
} from '@/types/release';

const initialState: ReleaseState = {
  releases: [],
  release: {} as NodeRelease,
  os: 'OTHER',
  bitness: '',
  platform: 'NVM',
};

const createDispatchActions = (
  dispatch: Dispatch<ReleaseAction>
): ReleaseDispatchActions => ({
  setVersion: payload => dispatch({ type: 'SET_VERSION', payload }),
  setOS: payload => dispatch({ type: 'SET_OS', payload }),
  setBitness: payload => dispatch({ type: 'SET_BITNESS', payload }),
  setPlatform: payload => dispatch({ type: 'SET_PLATFORM', payload }),
});

export const ReleaseContext = createContext<ReleaseContextType>({
  ...initialState,
  ...createDispatchActions(() => {}),
});

export const ReleaseProvider: FC<PropsWithChildren<ReleaseProviderProps>> = ({
  children,
  releases,
  initialRelease,
}) => {
  const getReleaseFromVersion = (version: string) =>
    releases.find(({ versionWithPrefix }) => versionWithPrefix === version) ??
    ({} as NodeRelease);

  const releaseReducer = (state: ReleaseState, action: ReleaseAction) => {
    switch (action.type) {
      case 'SET_VERSION':
        return { ...state, release: getReleaseFromVersion(action.payload) };
      case 'SET_OS':
        return { ...state, os: action.payload };
      case 'SET_BITNESS':
        return { ...state, bitness: action.payload };
      case 'SET_PLATFORM':
        return { ...state, platform: action.payload };
      default:
        return state;
    }
  };

  const [state, dispatch] = useReducer(releaseReducer, {
    ...initialState,
    releases: releases,
    release: initialRelease,
  });

  const actions = useMemo(() => createDispatchActions(dispatch), [dispatch]);

  return (
    <ReleaseContext.Provider value={{ ...state, ...actions }}>
      {children}
    </ReleaseContext.Provider>
  );
};
