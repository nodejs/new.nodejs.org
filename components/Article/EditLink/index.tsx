import { FaPencilAlt } from 'react-icons/fa';
import { FormattedMessage } from 'react-intl';
import styles from './index.module.scss';
import { useLocale } from './../../../hooks/useLocale';
import type { FC } from 'react';

type EditLinkProps = {
  absolutePath?: string;
  relativePath?: string;
  editPath?: string;
};

// TODO(HinataKah0): Change branch from major/website-redesign to main

const baseEditURL =
  'https://github.com/nodejs/nodejs.org/edit/major/website-redesign';

const translationReadmeURL =
  'https://github.com/nodejs/nodejs.org/blob/major/website-redesign/TRANSLATION.md';

const translationKeyPrefix = 'components.article.editLink.title';

type EditLinkParams = {
  translationKey: string;
  href: string;
};

const getEditLinkParams = (
  { absolutePath, relativePath, editPath }: EditLinkProps,
  lang: string
): EditLinkParams => {
  if (lang === 'en') {
    // Initial content development is done on GitHub in English
    return {
      translationKey: `${translationKeyPrefix}.edit`,
      href:
        absolutePath ||
        (relativePath
          ? `${baseEditURL}/pages/en/${relativePath}`
          : `${baseEditURL}/${editPath}`),
    };
  }

  return {
    translationKey: `${translationKeyPrefix}.translate`,
    href: translationReadmeURL,
  };
};

const EditLink: FC<EditLinkProps> = ({
  absolutePath,
  relativePath,
  editPath,
}) => {
  const { currentLocale } = useLocale();

  if (!relativePath && !editPath && !absolutePath) {
    return null;
  }

  const editLinkParams = getEditLinkParams(
    { absolutePath, relativePath, editPath },
    currentLocale.code
  );

  return (
    <div className={styles.edit}>
      <a href={editLinkParams.href}>
        <FormattedMessage id={editLinkParams.translationKey} tagName="span" />
        <FaPencilAlt />
      </a>
    </div>
  );
};

export default EditLink;
