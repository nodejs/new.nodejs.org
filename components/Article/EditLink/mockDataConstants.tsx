import type { AppProps } from '../../../types';

export const exampleAbsolutePath =
  'https://github.com/nodejs/nodejs.org/edit/major/website-redesign/pages/en/get-involved/contribute.md';
export const exampleRelativePath = 'get-involved/contribute.md';
export const exampleEditPath = 'pages/en/get-involved/contribute.md';

export const i18nMockDataEnglish = {
  currentLocale: {
    code: 'en',
  },
  localeMessages: {
    'components.article.editLink.title.edit':
      'Edit this page on GitHub (English)',
  },
} as unknown as AppProps['i18nData'];

export const i18nMockDataNonEnglish = {
  currentLocale: {
    code: 'xx',
  },
  localeMessages: {
    'components.article.editLink.title.translate':
      'Interested to help with translations? (Non-English)',
  },
} as unknown as AppProps['i18nData'];
