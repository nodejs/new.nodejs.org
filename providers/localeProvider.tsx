import { createContext } from 'react';
import { IntlProvider } from 'react-intl';
import i18nConfig from '../i18n/config.json';
import type { FC, PropsWithChildren } from 'react';
import type { AppProps, LocaleContext as LocaleContextType } from '../types';

// Retrieves all the Available Languages
const availableLocales = i18nConfig
  .filter(({ enabled }) => enabled)
  .map(({ code, localName, name }) => ({ code, localName, name }));

type LocaleProviderProps = PropsWithChildren<{
  i18nData: AppProps['i18nData'];
}>;

const fallbackData = { currentLocale: { code: 'en' } } as AppProps['i18nData'];

export const LocaleContext = createContext<LocaleContextType>(undefined as any);

export const LocaleProvider: FC<LocaleProviderProps> = ({
  i18nData,
  children,
}) => {
  const { currentLocale, localeMessages } = i18nData || fallbackData;

  return (
    <LocaleContext.Provider value={{ ...i18nData, availableLocales }}>
      <IntlProvider
        locale={currentLocale.code}
        messages={localeMessages}
        onError={() => null}
      >
        {children}
      </IntlProvider>
    </LocaleContext.Provider>
  );
};
