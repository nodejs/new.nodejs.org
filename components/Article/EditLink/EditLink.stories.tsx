import {
  exampleRelativePath,
  i18nMockDataEnglish,
  i18nMockDataNonEnglish,
} from './mockDataConstants';
import { LocaleProvider } from './../../../providers/localeProvider';
import EditLink from './index';
import type { Meta as MetaObj, StoryObj } from '@storybook/react';

type Story = StoryObj<typeof EditLink>;
type Meta = MetaObj<typeof EditLink>;

export const Edit: Story = {
  args: {
    relativePath: exampleRelativePath,
  },
  render: (args: any) => {
    return (
      <LocaleProvider i18nData={i18nMockDataEnglish}>
        <EditLink {...args} />
      </LocaleProvider>
    );
  },
};

export const Translate: Story = {
  render: () => {
    return (
      <LocaleProvider i18nData={i18nMockDataNonEnglish}>
        <EditLink relativePath={exampleRelativePath} />
      </LocaleProvider>
    );
  },
};

export const Empty: Story = {};

export default { component: EditLink } as Meta;
