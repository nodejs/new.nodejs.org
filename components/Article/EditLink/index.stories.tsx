import EditLink from './index';
import type { Meta as MetaObj, StoryObj } from '@storybook/react';

type Story = StoryObj<typeof EditLink>;
type Meta = MetaObj<typeof EditLink>;

export const Default: Story = {
  args: { relativePath: 'get-involved/contribute.md' },
};

export default { component: EditLink } as Meta;
