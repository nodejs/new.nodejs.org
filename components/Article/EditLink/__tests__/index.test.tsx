import { render, screen } from '@testing-library/react';
import EditLink from './../index';
import { LocaleProvider } from '../../../../providers/localeProvider';
import type { AppProps } from '../../../../types';

jest.mock('next/router', () => ({
  useRouter: jest.fn().mockImplementation(() => ({
    asPath: '',
  })),
}));

const absolutePath =
  'https://github.com/nodejs/nodejs.org/edit/major/website-redesign/pages/en/get-involved/contribute.md';
const relativePath = 'get-involved/contribute.md';
const editPath = 'pages/en/get-involved/contribute.md';

const i18nDataEditMode = {
  currentLocale: {
    code: 'en',
  },
  localeMessages: {
    'components.article.editLink.title.edit': 'Edit this page on GitHub',
  },
} as unknown as AppProps['i18nData'];

const i18nDataTranslateMode = {
  currentLocale: {
    code: 'xx',
  },
  localeMessages: {
    'components.article.editLink.title.translate':
      'Interested to help with translations?',
  },
} as unknown as AppProps['i18nData'];

describe('EditLink component', () => {
  it('edit mode renders correctly', () => {
    const { container } = render(
      <LocaleProvider i18nData={i18nDataEditMode}>
        <EditLink relativePath={relativePath} />
      </LocaleProvider>
    );
    expect(container).toMatchSnapshot();
  });

  it('translate mode renders correctly', () => {
    const { container } = render(
      <LocaleProvider i18nData={i18nDataTranslateMode}>
        <EditLink relativePath={relativePath} />
      </LocaleProvider>
    );
    expect(container).toMatchSnapshot();
  });

  it('renders without a relative path', () => {
    const { container } = render(
      <LocaleProvider i18nData={i18nDataEditMode}>
        <EditLink relativePath={undefined} />
      </LocaleProvider>
    );
    expect(container).toMatchSnapshot();
  });

  it('produces correct relative path', () => {
    render(
      <LocaleProvider i18nData={i18nDataEditMode}>
        <EditLink relativePath={relativePath} />
      </LocaleProvider>
    );
    expect(screen.getByRole('link')).toHaveAttribute('href', absolutePath);
  });

  it('produces correct edit path', () => {
    render(
      <LocaleProvider i18nData={i18nDataEditMode}>
        <EditLink editPath={editPath} />
      </LocaleProvider>
    );
    expect(screen.getByRole('link')).toHaveAttribute('href', absolutePath);
  });
});
