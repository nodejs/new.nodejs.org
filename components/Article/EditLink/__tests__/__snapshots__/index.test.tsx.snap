// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditLink component edit mode renders correctly 1`] = `
<div>
  <div
    class="edit"
  >
    <a
      href="https://github.com/nodejs/nodejs.org/edit/major/website-redesign/pages/en/get-involved/contribute.md"
    >
      <span>
        Edit this page on GitHub
      </span>
      <svg
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 512 512"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M497.9 142.1l-46.1 46.1c-4.7 4.7-12.3 4.7-17 0l-111-111c-4.7-4.7-4.7-12.3 0-17l46.1-46.1c18.7-18.7 49.1-18.7 67.9 0l60.1 60.1c18.8 18.7 18.8 49.1 0 67.9zM284.2 99.8L21.6 362.4.4 483.9c-2.9 16.4 11.4 30.6 27.8 27.8l121.5-21.3 262.6-262.6c4.7-4.7 4.7-12.3 0-17l-111-111c-4.8-4.7-12.4-4.7-17.1 0zM124.1 339.9c-5.5-5.5-5.5-14.3 0-19.8l154-154c5.5-5.5 14.3-5.5 19.8 0s5.5 14.3 0 19.8l-154 154c-5.5 5.5-14.3 5.5-19.8 0zM88 424h48v36.3l-64.5 11.3-31.1-31.1L51.7 376H88v48z"
        />
      </svg>
    </a>
  </div>
</div>
`;

exports[`EditLink component renders without a relative path 1`] = `<div />`;

exports[`EditLink component translate mode renders correctly 1`] = `
<div>
  <div
    class="edit"
  >
    <a
      href="https://github.com/nodejs/nodejs.org/blob/major/website-redesign/TRANSLATION.md"
    >
      <span>
        Interested to help with translations?
      </span>
      <svg
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 512 512"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M497.9 142.1l-46.1 46.1c-4.7 4.7-12.3 4.7-17 0l-111-111c-4.7-4.7-4.7-12.3 0-17l46.1-46.1c18.7-18.7 49.1-18.7 67.9 0l60.1 60.1c18.8 18.7 18.8 49.1 0 67.9zM284.2 99.8L21.6 362.4.4 483.9c-2.9 16.4 11.4 30.6 27.8 27.8l121.5-21.3 262.6-262.6c4.7-4.7 4.7-12.3 0-17l-111-111c-4.8-4.7-12.4-4.7-17.1 0zM124.1 339.9c-5.5-5.5-5.5-14.3 0-19.8l154-154c5.5-5.5 14.3-5.5 19.8 0s5.5 14.3 0 19.8l-154 154c-5.5 5.5-14.3 5.5-19.8 0zM88 424h48v36.3l-64.5 11.3-31.1-31.1L51.7 376H88v48z"
        />
      </svg>
    </a>
  </div>
</div>
`;
